File: 880-Decoded_String_at_Index/Decoded_String_at_Index.go
package decodedstringatindex

func decodeAtIndex(s string, k int) string {
	decodedLength := 0

	for i := range s {
		c := s[i]
		if isDigit(c) {
			decodedLength *= int(c - '0')
		} else {
			decodedLength++
		}
	}

	for i := len(s) - 1; i >= 0; i-- {
		c := s[i]
		k %= decodedLength
		if k == 0 && isDigit(c) == false {
			return string(c)
		}

		if isDigit(c) {
			decodedLength /= int(c - '0')
		} else {
			decodedLength--
		}
	}

	return ""

}

func isDigit(b byte) bool {
	if b >= '2' && b <= '9' {
		return true
	}

	return false
}

File: 0880-Decoded_String_at_Index/Decoded_String_at_Index.java
class Solution {
    public String decodeAtIndex(String s, int k) {
        long decodedLength = 0;
        
        // get the total length of the decoded text
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(isDigit(c)){
                decodedLength *= c - '0';
            }
            else {
                decodedLength++;
            }
        }
        
        // Work backward to find the character --> Dont have to store the string which might result to out of memory
        for(int i = s.length() - 1; i >= 0; i--){
            char c = s.charAt(i);
            k %= decodedLength;
            if(k == 0 && !isDigit(c))   // If k == 0, it means it is a the "correct" iteration, return the next non-digit character
                return "" + c;
            
            if(isDigit(c)){
                decodedLength /= c - '0';
            }
            else{
                decodedLength--;
            }
        }
        
        return "";
    }
    
    private boolean isDigit(char c){
        return c >= '2' && c <= '9';
    }
}

